packages/ove-asset-manager/src/OVE.Service.AssetManager/Controllers/OVEAssetModelController.cs

```diff
@@ -0,0 +1,517 @@
+        /// <param name="upload">optional new file</param>
+        /// <returns></returns>
+        [HttpPost]
+        [DisableRequestSizeLimit]
+        [Route("/OVEAssetModelController/Edit/{id}.{format?}")]
+        public async Task<ActionResult<OVEAssetModel>> Edit(string id, [Bind("Project,Name,Description,Id,StorageLocation,Processed,ProcessingState,Service,AssetMeta,LastModified")]
+            OVEAssetModel oveAssetModel,[FromForm] IFormFile upload) {
+            if (id != oveAssetModel.Id) {
+                return NotFound();
+            }
+
+            var oldAssetModel = await _context.AssetModels.FirstOrDefaultAsync(m => m.Id == id);
+            if (oldAssetModel == null) {
+                return NotFound();
+            }
+
+            // concurrent fields need to be updated by themselves and atomically. 
+            bool need2UpdateProcessingState = oldAssetModel.ProcessingState != oveAssetModel.ProcessingState;
+            if (need2UpdateProcessingState) { 
+                oveAssetModel.ProcessingState = oldAssetModel.ProcessingState;
+            }
+
+            if (!_serviceRepository.ValidateServiceChoice(oveAssetModel.Service, upload)) {
+                ModelState.AddModelError("Service", "Service does not support File Type");
+            }
+            
+            if (ModelState.IsValid) {
+                try {
+                    if (oldAssetModel.Project != oveAssetModel.Project) {
+                       await _fileOperations.MoveFile(oldAssetModel, oveAssetModel); // todo not implemented
+                    }
+                    //stop EF from tracking the old version so that it will allow you to update the new version
+                    _context.Entry(oldAssetModel).State = EntityState.Detached;
+                    
+                    if (upload != null && upload.Length > 0) {
+                        if (!await _fileOperations.DeleteFile(oveAssetModel)) {
+                            return UnprocessableEntity("unable to delete old file");
+                        }
+
+                        if (!await _fileOperations.SaveFile(oveAssetModel, upload)) {
+                            return UnprocessableEntity("unable to save new file");
+                        }
+                        need2UpdateProcessingState = true;
+                    }
+                    oveAssetModel.LastModified = DateTime.Now;
+                    _context.Update(oveAssetModel);
+
+                    await _context.SaveChangesAsync();
+
+                    if (need2UpdateProcessingState) {
+                         oveAssetModel.ProcessingState = 0;
+                        _context.Update(oveAssetModel);
+                        await _context.SaveChangesAsync();
+                    }
+
+                }
+                catch (DbUpdateConcurrencyException) {
+                    if (!_context.AssetModels.Any(e => e.Id == oveAssetModel.Id)) {
+                        return NotFound();

```

Comment:
Isn't this a 500 error? compared to a 404 error?

Reply:
404 is NotFound, that is eactly the situation we have here, the specified asset ID does not exist in the database. 
If thats not the case the exception will be thrown as a 500.

Reply:
I think you get a [DbUpdateConcurrencyException](https://msdn.microsoft.com/en-us/library/system.data.entity.infrastructure.dbupdateconcurrencyexception(v=vs.113).aspx) when more than one person tries to perform the same operation and not when something is not found. 

Now what you are trying to do here is probably deal with a situation where another update changed the table and this record does not exist as a result. 

But, it may very well be the case that you deleted the old file but got the exception when trying to save the new file. If so the `404 - NotFound` status is wrong AFAIU. WDYT?

Reply:
@davidbirchwork, precisely - so not a 404 I suppose? I don't think that would mean anything to the API user. It would be a server-side error (5xx) too, because the user could have been impacted by another user.

Reply:
the if test  on line 458 establishes that no asset in the database exists for some reason. Therefore you can't edit it. a 404 seems logical to me. 

have changed the throw to return a Conflicted 409 HTTP status.